common/environment/environment.cpp

```diff
@@ -2362,3 +2362,54 @@ extern ENVIRONMENT_API unsigned long readSizeSetting(const char * sizeStr, const
     }
     return size;
 }
+
+unsigned getAccessibleServiceURLList(const char *serviceType, std::vector<std::string> &list)
+{
+    unsigned added = 0;
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true);
+    Owned<IConstEnvironment> daliEnv = factory->openEnvironment();
+    Owned<IPropertyTree> env = &daliEnv->getPTree();
+    if (env.get())
+    {
+        StringBuffer fileMetaServiceUrl;
+        StringBuffer espInstanceComputerName;
+        StringBuffer bindingProtocol;
+        StringBuffer xpath;
+        StringBuffer instanceAddress;
+        StringBuffer espServiceType;
+
+        Owned<IPropertyTreeIterator> espProcessIter = env->getElements("Software/EspProcess");
+        ForEach(*espProcessIter)
+        {
+            Owned<IPropertyTreeIterator> espBindingIter = espProcessIter->query().getElements("EspBinding");
+            ForEach(*espBindingIter)
+            {
+                xpath.setf("Software/EspService[@name=\"%s\"]/Properties/@type",  espBindingIter->query().queryProp("@service"));
+
+                if (strisame(env->queryProp(xpath), serviceType))
+                {
+                    if (espBindingIter->query().getProp("@protocol", bindingProtocol.clear()))
+                    {
+                        Owned<IPropertyTreeIterator> espInstanceIter = espProcessIter->query().getElements("Instance");
+                        ForEach(*espInstanceIter)
+                        {
+                            if (espInstanceIter->query().getProp("@computer", espInstanceComputerName.clear()))
+                            {
+                                xpath.setf("Hardware/Computer[@name=\"%s\"]/@netAddress", espInstanceComputerName.str());
+                                if (env->getProp(xpath.str(), instanceAddress.clear()))
+                                {
+                                    fileMetaServiceUrl.setf("%s://%s:%d", bindingProtocol.str(), instanceAddress.str(), espBindingIter->query().getPropInt("@port",8010));
```

Comment:
If we're trying to find all availble eclwatch instances, looking for 8010 might be ok, but there's nothing keeping an admin from assigning to another port and btw 18010 is the default for https. We might be better off looking for bound EspServices of type=WsSCM or plugin= ws_smc.
If we're looking for wsdfu, we might want to search for bound EspServices of type="WsDfu" and/or plugin="ws_dfu"

Reply:
Agreed, this code was essentially a tidied up copy of code that does the same thing in fileservices.cpp.
Which also is hard-wired to look for port 8010 - that's ancient code afaik - but also wrong for the same reasons.

The code in fileservices.cpp should use a/this common method as well, then at least there's 1 place to fix this issue - but it's not really a new issue and given fileservices has been living with it for years, I think we can think about how best to do it for all such client discovering services in one place (like environment.*) and fix it once.
Let's open a new JIRA for this issue.

Reply:
ok opened this: https://track.hpccsystems.com/browse/HPCC-21636